src/t_zset.c

```diff
@@ -2023,7 +2022,6 @@ void zremrangeGenericCommand(client *c, zrange_type rangetype) {
             deleted = zslDeleteRangeByLex(zs->zsl,&lexrange,zs->dict);
             break;
         }
-        if (htNeedsResize(zs->dict)) dictResize(zs->dict);
```

Comment:
When removing a lot of elements in commands like ZREMRANGE, let's say we remove 95% of the elements in a dict of 1000 elements, then after this change the dict will shink to 10% (100 buckets). Before this change, it would shrink to 5% (50 buckets), right?

If this was a useful behaviour that we want to keep, then we could temporarily disable automatic resize on the dict before we call zslDeleteRangeByScore and enable it afterwards, which would trigger a resize to 5%, to preserve the old behaviour, only in a few cases likely to remove many elements such as ZREMRANGE. (I don't think it's very useful for ZREM, for example.) WDYT?

Reply:
good point.
so some `startMassRemoval()` and `endMassRemoval()`?

Reply:
There is another case: if a dictionary is in the process of rehashing and there are large batches of insertions or deletions happening simultaneously, the target size for that rehash may no longer be appropriate. Perhaps it would be worth considering an additional check at the end of the rehash process to determine if another rehash is necessary? This approach might seem a bit complex, not sure if it would be suitable.

Reply:
> so some `startMassRemoval()` and `endMassRemoval()`?

I was thinking something like `dictSetAllowResize()`

Reply:
obviously it'll already be too late to stop the current re-hash.
why is a check at the end needed? wouldn't it be handled by the cron normally sooner or later?

Reply:
>wouldn't it be handled by the cron normally sooner or later?

the main dict can be handled by cron, but the other data structures like hash and zset cannot.

Reply:
so we try to implicitly trigger resize when the dict is modified (expand in dictAdd, and shrink in dictDelete)
we do incremental rehashing in dictFind, and if that incremental re-hash results in the resize being completed, we consider another resize.

but the slow and fast cron scan is still need to handle a completely stale database (i.e. when traffic to a db / dict completely stopped, even read-only one).
and for dict based keys, we don't have that.
i wonder how important is this case (no traffic to a certain dict), and if we can align the main dict with what happens for a hash key, and eliminate this cron resize job...

i suppose we can't, but then i wonder how important is it to fix this problem (for hash keys that need an additional re-hashing), when they do still receive read traffic, but don't receive write traffic. 

i'm leaning towards putting that aside.

Reply:
@zuiderkwast Thanks for your advice. I add a variable named `pauseAutoResize` in dict to control whether the dict can resize automatically. Its usage is similar to `pauserehash`. 

> i suppose we can't, but then i wonder how important is it to fix this problem (for hash keys that need an additional re-hashing), when they do still receive read traffic, but don't receive write traffic.

In `dictFind`, we perform a step of rehashing but can‘t trigger a rehash. Although `dictFind` doesn't change the size of the dict, the dict may meet the condition of expanding or shrinking. For example, if users insert many entries into dict during bgsave and only read it after bgsave, the size of dict may bigger than the number of buckets but smaller than `dict_force_resize_ratio` times of the number of buckets. In this case, the dict won't be rehashed.  I think we may be have a try.